### PR TITLE
#165965300 Social authentication refactor

### PR DESCRIPTION
**Pull request by JamesMudidi**
#### What does this PR do?
- Refactor Social Authentication

#### Description of Task to be completed?
- Refactor the authentication password to contain a random password

#### How should this be manually tested?
After cloning the repository,

- Navigate into it and install the dependencies using `pip3 install -r requirements.txt`
- Setup a database with PostgreSQL
- Set up your environment variables following the sample.env file provided and then run `source .env`
- Make migrations using `python manage.py makemigrations`
- Appl the migrations using `python manage.py migrate`
- Run the application using `python3 manage.py runserver`
- Create a developer account on any of the social media platforms supported
- Start postman and access the desired endpoint using `http://127.0.0.1:8000/api/auth/<social platform>/`. where <social platform> can be facebook
- Generate and copy the token from the already created developer account of the desired social platform eg. facebook
- use the token as shown in the screenshot below

#### Any background context you want to provide?
- After doing more research on the social auth feature, we decided to have a random password for the users
- This feature works with third-party Apis so you need to have some knowledge on the use of those Apis before proceeding to test the feature. The links below will help you get started with creating an account to become a developer for Facebook and Google where you can generate tokens and test the API.
- [Facebook](https://developers.facebook.com/)
- [Google](https://developers.google.com/oauthplayground/)

#### Screenshots

Facebook Screenshoot
Screenshot 2019-05-07 at 12 11 40 PM

Google ScreenShoot
Screenshot 2019-05-07 at 12 13 15 PM

#### What are the relevant pivotal tracker stories?
[165965300](https://www.pivotaltracker.com/story/show/165965300)

### CODE REVIEW

File: authors/apps/authentication/social.py [hunk 1]
```diff
@@ -8,21 +8,23 @@
 
 def register_social_user(email, username):
     user = User.objects.filter(email=email)
+    password = User.objects.make_random_password()
```
Comment (nadralia) [inline authors/apps/authentication/social.py]: Nice Work @JamesMudidi could you specify the length/allowed chars of the password like
`passwrod = User.objects.make_random_password(length=14, allowed_chars="abcdefghjkmnpqrstuvwxyz01234567889") `# zvk0hawf8m6394,

**Review by nadralia — APPROVED** [review]

**Review by llwasampijja — APPROVED** [review]
Tested the feature and works well on heroku, nice work @JamesMudidi

File: authors/apps/authentication/social.py [hunk 2]
```diff
@@ -1,33 +1,41 @@
...
+    and user login
+    '''
     user = User.objects.filter(email=email)
+    password = User.objects.make_random_password(length=20)
 
     if user.exists():
         current_user = User.objects.get(email=email)
+        current_user.set_password(password)
         current_user.is_active = True
         current_user.save()
```
Comment (bisonlou) [inline authors/apps/authentication/social.py]: Great work @JamesMudidi. However, I wonder why we have to re-save the user if he/she already exists?

**Review by bisonlou — DISMISSED** [review]
Well done

File: authors/apps/authentication/social.py [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
 def register_social_user(email, username):
+    '''
+    This function handles user registration
+    and user login
+    '''
     user = User.objects.filter(email=email)
+    password = User.objects.make_random_password(length=20)
 
     if user.exists():
         current_user = User.objects.get(email=email)
```
Comment (kalsmic) [inline authors/apps/authentication/social.py:20]: @JamesMudidi, on registration, if the user already exists simply get the token from the user object.
```
token = current_user.token
return token
```
Replace lines 20 to 25

**Review by kalsmic — CHANGES_REQUESTED** [review]

**Review by kalsmic — APPROVED** [review]

**Pull request merged**